Offline Pages : Added Support for unpublished revision to Page List

jd-alexander opened the pull request:
Fixes #11131 

## Findings

The `Page List` didn't support loading an unpublished revision when changes were made on other devices. 

## Solution
* To reuse logic that allows the `Post List` to support this functionality. Most of this logic lives within the `PostListDialogHelper`. Since it has way more responsibilities than required in this PR, a `PageListDialogHelper` was created to extract only the functionality needed. 

* While adding this feature, I noticed that the `Delete Page Confirmation Dialog` was utilizing the `BasicFragmentDialog` and mixing both implementations would reduce code quality so I opted to migrate the `Delete Confirmation Dialog` to the `PageListDialogHelper` for consistency. 

* For the "You've made unsaved changes to this page" label, I didn't add it since @malinajirka will be working on a PR to refactor how labels are generated and that would cause a conflict. 

* I reused the tracking events from the `PostListDialogHelper` but I am wondering if pages need to be tracked separately from posts.  

## Testing

### Page Auto Revision Testing 
1. Edit a page in WP Calypso and don't click save. You can monitor the network tab in the chrome debugging tools to see when an `autosave` request is made since this implementation doesn't contain a label. 
2. Ensure the device is online and then refresh the page. 
3. Click on the page that was edited online and you should see the dialog below. 

Before Click | After Click
--------|-------
<img src="https://user-images.githubusercontent.com/1509205/73781137-c1f7e200-475d-11ea-90b2-566fc1aefbd8.png" width="320"> |   <img src="https://user-images.githubusercontent.com/1509205/73781152-c7552c80-475d-11ea-984c-9113106aec72.png" width="320">

4. Verify that clicking `The Version From Another Device` shows the contents that exist in Calypso. 
5. Click back without saving and press the page again to verify that `The Version From This App` is accurate.

### Page Deletion Testing 
1. Go to the `Trash` tab and click the `More` icon. 
2. Select `Delete Permanently` 
3. A dialog should be shown that asks you if you are sure you want to delete the page. 
4. Once you confirm the deletion, the page should be removed and the list should be empty. A snackbar should also be shown confirming deletion. 

Before | After 
--------|-------
<img src="https://user-images.githubusercontent.com/1509205/73781678-b2c56400-475e-11ea-82bd-4a98b083faca.png" width="320">        |    <img src="https://user-images.githubusercontent.com/1509205/73781731-c40e7080-475e-11ea-8ab5-aabb2cd38d38.png" width="320">

## Reviewing 
Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [x] I have considered adding accessibility improvements for my changes.
- [x] If it's feasible, I have added unit tests.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt:
```diff
@@ -444,14 +458,14 @@ class PagesViewModel
-
     fun onItemTapped(pageItem: Page) {
-        _editPage.postValue(pageMap[pageItem.id])
+        val page = postStore.getPostByRemotePostId(pageItem.id, site)
```
We might want to consider adding a incode TODO here and link https://github.com/wordpress-mobile/WordPress-Android/issues/11233. Wdyt?

jd-alexander replied:
Resolved in 155b6dd

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/ui/pages/PagesActivity.kt:
```diff
@@ -53,16 +53,16 @@ class PagesActivity : AppCompatActivity(),
+        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
+        if (fragment is PagesFragment) {
+            fragment.onPositiveClickedForBasicDialog(instanceTag)
+        }
```
Resolved in 28a76e4

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/ui/pages/PagesActivity.kt:
```diff
@@ -53,16 +53,16 @@ class PagesActivity : AppCompatActivity(),
         if (fragment is PagesFragment) {
-            fragment.onPageDeleteConfirmed(remoteId)
+            fragment.onNegativeClickedForBasicDialog(instanceTag)
         }
```
Resolved in 28a76e4

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostUtils.java:
```diff
@@ -460,8 +460,13 @@ public static String getConflictedPostCustomStringForDialog(PostModel post) {
+        return getCustomStringForAutosaveRevisionDialog(post, false);
+    }
+
+    public static UiStringText getCustomStringForAutosaveRevisionDialog(PostModel post, boolean isPage) {
```
Resolved in aa7d8db 8dd300b

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListDialogHelper.kt:
```diff
@@ -0,0 +1,89 @@
+
+    fun showDeletePageConfirmationDialog(pageId: RemoteId, pageTitle: String) {
+        val dialogMessage = {
+            WordPress.getContext().getString(R.string.page_delete_dialog_message, pageTitle)
```
> PageListDialogHelper is part of the viewmodel layer. I believe it shouldn't have access to the context. That's basically the main reason why we use UiString wrappers in this class. We might want to consider extending UIString so it supports parameters, wdyt?

Thanks for this! Modifying the `UiHelper` is a simple sensible approach. 

> I'm also not sure I follow the lambda approach. Why do we need to create a lambda just to invoke it once a few lines below?

Indeed, I was trying to avoid the long line of code within the `dialogHolder`. The solution above makes this cleaner. Thanks! 

Resolved in d5f27b7 8dd300b d276c95

malinajirka replied:
>Indeed, I was trying to avoid the long line of code within the dialogHolder.

I understand that part, but why we need the lambda. Wouldn't simply calling ```val dialogMessage = WordPress.getContext().getString(R.string.page_delete_dialog_message, pageTitle)``` without the parentheses be enough? If it was just a mistake, don't worry about it. I was just curious if it's a trick I'm not aware of.

>Resolved in d522cd2 d5f27b7 d276c95
In the above solution, I am passing the params as a List and then converting it to an array d5f27b7 when I realized it didn't behave as intended. Is this okay to do? I was going to use an array instead of a List but it's recommended that I implement equals & hashCode what do you think about this?

Using `Any` might bite us in the future. AFAIU  the params still need to be strings, so I'd try using `UIString`. Moreover, if someone passes UIString there, the current implementation will print something like `@ad87adf9`.
We might want to consider using [varargs](https://kotlinlang.org/docs/reference/functions.html#variable-number-of-arguments-varargs) instead of passing a list and converting it into a typed array. 

Wdyt?

jd-alexander replied:
> If it was just a mistake, don't worry about it. I was just curious if it's a trick I'm not aware of.

Yes. It wasn't anything you weren't aware of 😄 

> Using Any might bite us in the future. AFAIU the params still need to be strings, so I'd try using UIString. Moreover, if someone passes UIString there, the current implementation will print something like @ad87adf9.

Thanks for expounding on this. I am going to utilize the `UiString` so that the params will always be strings. You know what I was thinking, I normally passed in different types so I was thinking about the support from that angle but now I am realizing those are still converted to strings with formatting. 

> We might want to consider using varargs instead of passing a list and converting it into a typed array.
Wdyt? 

Good point ! I considered doing this at first but what I realized was `varargs` [aren't supported in the constructor](https://stackoverflow.com/questions/31415093/constructor-call-in-kotlin-using-vararg). Let me know if there's another way to do this.

Also, I was going to pass it as an array but I end up using the `map` operator in the solution to transform each `UiString` to a string using the `UiHelper` method so that converts it back to a list anyway and it ends up being converted to a typed array so that the spread operator can be used. If you see a way this can be further simplified, let me know 😉 Resolved in 4b912f0

malinajirka replied:
> I considered doing this at first but what I realized was varargs aren't supported in the constructor. Let me know if there's another way to do this.

Ohh, I didn't know that. Thanks for clarification! I think using list is ok, especially since as you mentioned we use the `map`.

It all looks good now, I'm just curious. What is the `*` good for in `*uiString.params.map ...` ? I have never seen such usage. I know the usage when specifying generic type eg `<*>`, but I'm not sure what it does in front of a statement.

malinajirka replied:
TIL about the `*` - Destructuring operator. I understand the usage now. LGTM ;)!

jd-alexander replied:
Awesome 😉

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListDialogHelper.kt:
```diff
@@ -0,0 +1,89 @@
+            CONFIRM_DELETE_PAGE_DIALOG_TAG -> pageIdForDeleteDialog?.let {
+                pageIdForDeleteDialog = null
+                deletePage(it)
+            }
```
Resolved in ca2ccdc

malinajirka replied:
Is there any advantage using `run` instead of simply calling ```?: throw NullPointerException("pageIdForAutosaveRevisionResolutionDialog shouldn't be null.")```?

jd-alexander replied:
Nope! Thanks! I sometimes forget these syntax/language tricks 😅  Resolved in 2a6afe5

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListDialogHelper.kt:
```diff
@@ -0,0 +1,89 @@
+                pageIdForAutosaveRevisionResolutionDialog = null
+                editPage(it, true)
+                analyticsTracker.track(UNPUBLISHED_REVISION_DIALOG_LOAD_UNPUBLISHED_VERSION_CLICKED)
+            }
```
Resolved in ca2ccdc

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListDialogHelper.kt:
```diff
@@ -0,0 +1,89 @@
+                // open the editor with the local page (don't use the auto save version)
+                editPage(it, false)
+                analyticsTracker.track(UNPUBLISHED_REVISION_DIALOG_LOAD_LOCAL_VERSION_CLICKED)
+            }
```
Resolved in ca2ccdc

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListDialogHelper.kt:
```diff
@@ -0,0 +1,89 @@
+private const val CONFIRM_ON_AUTOSAVE_REVISION_DIALOG_TAG = "CONFIRM_ON_AUTOSAVE_REVISION_DIALOG_TAG"
+private const val CONFIRM_DELETE_PAGE_DIALOG_TAG = "CONFIRM_DELETE_PAGE_DIALOG_TAG"
+
+class PageListDialogHelper(
```
Thanks. This seems like a good thing to do. I am going to create an issue for this so it can be done separately in the code cleanup. https://github.com/wordpress-mobile/WordPress-Android/issues/11272

jd-alexander commented:
> Great job @jd-alexander! I haven't found any major issues. I've left a couple of minor comments with some possible improvements. Let me know what you think ;)

Awesome. Thanks again for another thoughtful review :bowtie: 

> We might want to consider extracting each dialog into its own class. That'd allow us to share the logic between the screens. But let's consider this during the cleanup phase of the project.

Indeed! I created an issue to track this https://github.com/wordpress-mobile/WordPress-Android/issues/11272

> Good call ;)! It might be even better to separate it into a different PR, but it doesn't matter since this PR is pretty straightforward. Thanks!

Thanks for the comment 😄 

> Hmm, we might want to consider adding something like "type:post/page" attribute to the event. Wdyt?

Makes sense! I added it ac2129b So I should create an issue to add this to the `PostDialogHelper` right?

> Btw I noticed branch names on your PRs don't follow the issue/[issue_number]_more_details pattern. AFAIK our CI doesn't count on the pattern, but I think it'd be good to follow it for consistency with other branch names.

Thanks! I sometimes create a quick branch when I start playing around, renamed it locally and didn't realize I didn't push the new name. I will make sure all new PRs adhere to this from the start for sure 👍

jkmassel commented:
We're freezing `14.2` today, so this PR is being bumped to `14.3`. If you'd like it to ship with `14.2`, please merge it into `release/14.2` and ping me – I'll be happy to cut a new beta release.

malinajirka commented:
>Makes sense! I added it ac2129b So I should create an issue to add this to the PostDialogHelper right?

Thanks! Yeah, please create a ticket and link this usage there. I think we'll want share POST_TYPE to make sure no-one changes one place but forgets about the other.


Thank you for all the changes! It all LGTM. I'm merging this PR even though the connect tests are failing, since they are failing on a lot of the PRs. The issue is not related to this PR.
